ethereum/contracts/bridge/L1ERC20Bridge.sol

```diff
@@ -77,57 +113,154 @@ contract L1ERC20Bridge is IL1Bridge, IL1BridgeLegacy, AllowListed, ReentrancyGua
     /// @notice At the time of the function call, it is not yet deployed in L2, but knowledge of its address
     /// @notice is necessary for determining L2 token address by L1 address, see `l2TokenAddress(address)` function
     /// @param _governor Address which can change L2 token implementation and upgrade the bridge
-    /// @param _deployBridgeImplementationFee How much of the sent value should be allocated to deploying the L2 bridge
     /// implementation
-    /// @param _deployBridgeProxyFee How much of the sent value should be allocated to deploying the L2 bridge proxy
-    function initialize(
+    function initializeV2(
         bytes[] calldata _factoryDeps,
         address _l2TokenBeacon,
+        address _l2Bridge,
         address _governor,
-        uint256 _deployBridgeImplementationFee,
-        uint256 _deployBridgeProxyFee
-    ) external payable reentrancyGuardInitializer {
+        address _l2Governor
+    ) external payable reinitializer(2) {
         require(_l2TokenBeacon != address(0), "nf");
         require(_governor != address(0), "nh");
         // We are expecting to see the exact three bytecodes that are needed to initialize the bridge
         require(_factoryDeps.length == 3, "mk");
         // The caller miscalculated deploy transactions fees
-        require(msg.value == _deployBridgeImplementationFee + _deployBridgeProxyFee, "fee");
         l2TokenProxyBytecodeHash = L2ContractHelper.hashL2Bytecode(_factoryDeps[2]);
-        l2TokenBeacon = _l2TokenBeacon;
+        l2TokenBeaconStandardAddress = _l2TokenBeacon;
+        l2BridgeStandardAddress = _l2Bridge;
+        governor = _governor;
+        l2Governor = _l2Governor;
+
+        factoryDepsHash = keccak256(abi.encode(_factoryDeps));
+    }
+
+    function l2Bridge() external view returns (address) {
+        return l2BridgeAddress[eraChainId];
+    }
+
+    /// @notice Checks that the message sender is the governor
+    modifier onlyGovernor() {
```

Comment:
Just in case, L1ERC20Bridge is deployed as a Transparent Upgradable Proxy. In other words, governor of the proxy can not call the implementation. So the implementation will have to have a different governor